tests/test_components/test_heat_charge.py

```diff
@@ -714,3 +792,68 @@ def test_sim_data():
 
     with pytest.raises(pd.ValidationError):
         _ = heat_sim_data.updated_copy(simulation=sim)
+
+
+def test_unsteady_spec():
+    """Make sure the unsteady spec is working appropriately."""
+
+    heat_sim = make_heat_charge_heat_sim()
+    cond_sim = make_heat_charge_cond_sim()
+
+    time_spec = td.UnsteadySpec(time_step=1, total_time_steps=2, initial_temperature=1)
+
+    # normal case, shouldn't raise error
+    _ = heat_sim.updated_copy(time_spec=time_spec)
+
+    # negative time-step
+    with pytest.raises(pd.ValidationError):
+        _ = time_spec.updated_copy(time_step=-0.1)
+
+    # negative total number of time-steps
+    with pytest.raises(pd.ValidationError):
+        _ = time_spec.updated_copy(total_time_steps=-1)
+
+    # EQS
+    time_spec2 = time_spec.updated_copy(initial_voltage=2, initial_temperature=None)
+    _ = cond_sim.updated_copy(time_spec=time_spec2)
+
+    # check with mixed simulation
+    bc = list(heat_sim.boundary_spec)
+    bc.extend(cond_sim.boundary_spec)
+    structures = list(heat_sim.structures)
+    structures.extend(list(cond_sim.structures))
+    heat_sim2 = heat_sim.updated_copy(structures=list(set(structures)), boundary_spec=bc)
+    _ = heat_sim2.updated_copy(time_spec=time_spec)
+
+
+def test_charge_simulation():
+    """Make sure charge simulation produces the correct errors when needed."""
+
+    sim = td.HeatChargeSimulation(
+        structures=[CHARGE_SIMULATION.oxide, CHARGE_SIMULATION.p_side, CHARGE_SIMULATION.n_side],
+        medium=td.Medium(heat_spec=td.FluidSpec(), name="air"),
+        monitors=[CHARGE_SIMULATION.charge_global_mnt],
+        center=(0, 0, 0),
+        size=CHARGE_SIMULATION.sim_size,
+        grid_spec=td.UniformUnstructuredGrid(dl=0.05),
+        boundary_spec=[CHARGE_SIMULATION.bc_n, CHARGE_SIMULATION.bc_p],
+        devsim_settings=[CHARGE_SIMULATION.devsim_settings],
+    )
+
+    # at least one ChargeSimulationMonitor should be added
+    with pytest.raises(pd.ValidationError):
+        _ = sim.updated_copy(monitors=[])
+
+    # at least 2 VoltageBCs should be defined
+    with pytest.raises(pd.ValidationError):
+        _ = sim.updated_copy(boundary_spec=[CHARGE_SIMULATION.bc_n])
+
+    # define ChargeSimulation with no Semiconductor materials
+    medium = td.Medium(
+        electric_spec=td.ConductorSpec(permittivity=1, conductivity=1),
+        name="medium",
+    )
+    new_structures = [struct.updated_copy(medium=medium) for struct in sim.structures]
+
+    with pytest.raises(pd.ValidationError):
+        _ = sim.updated_copy(structures=new_structures)
```

Comment:
yeah that's true the validation error would catch most of these, I'm just thinking for user-experience user-interface testing completeness if these include all the possible ways that a user would create a HeatChargeSimulation? They'd still raise validationerrors if they're wrong, but I feel we want to test all instantiation flows both on the components and the preliminary creation methods/classes (such as the copy you have)

Reply:
But, isn't it the same? I mean, `updated_copy` or creating a new instance with all parameters? Or am I missing something?

Reply:
It should be the same, our `copy` function (called by `updated_copy`) explicitly calls `validate`. https://github.com/flexcompute/tidy3d/blob/0ff0e60bc7b02569f326d0a08ea04aa167e21ad1/tidy3d/components/base.py#L223

Maybe @daquinteroflex's question is a bit more philosophical again and not specific to this PR, since we use this way of testing (`updated_copy`) of many other validators too. As far as I know there is no easy way for the user to avoid the full validation when initializing a model. They could work somewhat hard if they want to do it, but in that case they should always explicitly realize that they're skipping validation. Actually we made that easier in pre/2.8, where now there's a `validate = True` argument passed to `updated_copy`. So it's possible to turn it off (which we can do internally for speed if needed), but again, it's very explicit.